### DISCUSSION TITLE
Unmap illuminate `[[`/`]]` and use the one from vimtex.

### DISCUSSION THREAD

**Discussion by Zeta611** [LazyVim/LazyVim#2048] [Q&A] [answered]
I'm trying to unmap `[[` and `]]` from illuminate just for TeX files. My attempt is to create a custom setup in `lua/plugins/illuminate.lua`:
```lua
return {
  "RRethy/vim-illuminate",
  event = "LazyFile",
  opts = {
    delay = 200,
    large_file_cutoff = 2000,
    large_file_overrides = {
      providers = { "lsp" },
    },
  },
  config = function(_, opts)
    require("illuminate").configure(opts)

    local function map(key, dir, buffer)
      vim.keymap.set("n", key, function()
        require("illuminate")["goto_" .. dir .. "_reference"](false)
      end, { desc = dir:sub(1, 1):upper() .. dir:sub(2) .. " Reference", buffer = buffer })
    end

    if vim.bo.filetype ~= "tex" then
      map("]]", "next")
      map("[[", "prev")
    end

    -- also set it after loading ftplugins, since a lot overwrite [[ and ]]
    vim.api.nvim_create_autocmd("FileType", {
      callback = function()
        if vim.bo.filetype ~= "tex" then
          return
        end
        local buffer = vim.api.nvim_get_current_buf()
        map("]]", "next", buffer)
        map("[[", "prev", buffer)
      end,
    })
  end,
  --  keys = {
  --    { "]]", desc = "Next Reference" },
  --    { "[[", desc = "Prev Reference" },
  --  },
}
```
from which I copied from the default.

**Comment by dpetka2001** [comment, marked as answer]
Create `~/.config/nvim/after/ftplugin/tex.lua` with the following contents
```lua
local buffer = vim.api.nvim_get_current_buf()
vim.schedule(function()
  vim.keymap.set("n", "]]", "<plug>(vimtex-]])", { buffer = buffer })
  vim.keymap.set("n", "[[", "<plug>(vimtex-[[)", { buffer = buffer })
end)
```
Test it to see if it works out the way you want it to.

**Reply by Zeta611** [reply]
Works great! Yet I think this is kind of an ad-hoc solution; I believe there should be a way to customize this at the level of LazyVim config.

**Reply by dpetka2001** [reply]
I'm sure it can be done another way, but the autocmds that I tried both in `illuminate`'s spec and `autocmds.lua` file, didn't seem to work. Unfortunately, I'm not that adept at Lua and LazyVim's internals myself to know the correct way of loading autocmds, so I hope someone else can come up with a better solution.

**Reply by Zeta611** [reply]
Still, it does solve my issue and thanks for the help :)

**Reply by PatrickJosh** [reply]
Unfortunately, this does not seem to work any more. I've not yet succeeded in replicating the behaviour using a different method.

I'd be grateful if someone has a working solution/an idea how to to it!

**Reply by dpetka2001** [reply]
Try in file `~/.config/nvim/lua/config/autocmds.lua` the following
```lua
vim.api.nvim_create_autocmd("LspAttach", {
  callback = function(ev)
    vim.schedule(function()
      if vim.bo[ev.buf].filetype == "tex" then
        vim.keymap.set("n", "]]", "<plug>(vimtex-]])", { buffer = ev.buf })
        vim.keymap.set("n", "[[", "<plug>(vimtex-[[)", { buffer = ev.buf })
      end
    end)
  end,
})
```

**Reply by PatrickJosh** [reply]
Hi, thank you for your quick response! 

I've tested it, and unfortunately, it does not seem to work. I've added a print statement to the innermost if-statement and I can see the message being printed, so the code is executed, but `[[` and `]]` still do not map to the vimtex commands.

**Reply by dpetka2001** [reply]
Hey, sorry was outside and responding from my phone. Looked into LazyVim code where these mappings get defined and the following should work I believe
```lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      texlab = {
        keys = {
          { "]]", "<plug>(vimtex-]])" },
          { "[[", "<plug>(vimtex-[[)" },
        },
      },
    },
  },
}
```
No need for the previous autocmd. Please check it out.

**Reply by dpetka2001** [reply]
Actually, you will need to follow latest HEAD instead of LazyVim stable release, because a fix for the LSP mappings has been merged but is not available in the latest release yet, only on main.

So, the following should work
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        texlab = {
          keys = {
            { "]]", "<plug>(vimtex-]])" },
            { "[[", "<plug>(vimtex-[[)" },
          },
        },
      },
    },
  },

  {
    "LazyVim/LazyVim",
    version = false,
  },
}
```
Don't forget to also update LazyVim from Lazy UI. Press `<leader>l` to show (it's lowercase L) and then `U` to upate all plugins. Also restart Neovim after all updates.

**Reply by PatrickJosh** [reply]
I've just tried it and it works. And since a new LazyVim version has just been released, I didn't even need to change the config to upgrade to main.

Thank you so much for your help!
